File: src/main/java/dev/alexnader/framed/util/BlockStateMap.java
package dev.alexnader.framed.util;

import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
import net.minecraft.block.BlockState;
import net.minecraft.state.property.Property;

import javax.annotation.Nullable;
import java.util.Set;

public class BlockStateMap<V> {
    private final Set<Property<?>> importantProperties;
    private final Int2ObjectOpenHashMap<V> states = new Int2ObjectOpenHashMap<>();

    private int hash(BlockState state) {
        int hash = 17;
        for (Property<?> property : importantProperties) {
            hash = 31 * hash + state.get(property).hashCode();
        }
        return hash;
    }

    public BlockStateMap(Set<Property<?>> importantProperties) {
        this.importantProperties = importantProperties;
    }

    public void put(BlockState state, V value) {
        states.put(hash(state), value);
    }

    public void put(int hash, V value) {
        states.put(hash, value);
    }

    public @Nullable V get(BlockState state) {
        return states.get(hash(state));
    }

    public @Nullable V get(int hash) {
        return states.get(hash);
    }

    public void clear() {
        states.clear();
    }

    public boolean containsEquivalent(BlockState state) {
        return states.containsKey(hash(state));
    }
}

File: src/main/kotlin/dev/alexnader/framity/util/BlockStateMap.kt
package dev.alexnader.framity.util

import net.minecraft.block.BlockState
import net.minecraft.state.property.Property

/**
 * Map from [BlockState] to [V]. Based on hash of combined properties.
 * Which properties to use are specified via [includeProperty].
 */
class BlockStateMap<V> : MutableMap<BlockState, V> {
    private val valuesMap: MutableMap<Int, V> = mutableMapOf()
    private val keysMap: MutableMap<Int, BlockState> = mutableMapOf()
    private val properties: MutableSet<Property<Comparable<Any>>> = mutableSetOf()

    override val size get() = valuesMap.size

    @Suppress("UNCHECKED_CAST")
    fun <P : Comparable<P>>includeProperty(prop: Property<P>) {
        this.properties.add(prop as Property<Comparable<Any>>)
    }

    /**
     * Combines the hashes of all properties from [properties] of [state].
     */
    private fun hash(state: BlockState): Int {
        return this.properties.map { state.get<Comparable<Any>>(it) }.hashCode()
    }

    override fun containsKey(key: BlockState) = keysMap.containsKey(hash(key))

    override fun containsValue(value: V) = valuesMap.containsValue(value)

    override fun get(key: BlockState) = valuesMap[hash(key)]

    override fun isEmpty() = valuesMap.isEmpty()

    override val entries: MutableSet<MutableMap.MutableEntry<BlockState, V>> = keysMap.entries.zip(valuesMap.entries).map { (k, v) -> Entry(k.value, v.value) as MutableMap.MutableEntry<BlockState, V> }.toMutableSet()

    override val keys: MutableSet<BlockState> = keysMap.values.toMutableSet()
    override val values: MutableCollection<V> = valuesMap.values

    override fun clear() = valuesMap.clear()

    override fun put(key: BlockState, value: V): V? {
        val hash = hash(key)
        keysMap[hash] = key
        return valuesMap.put(hash, value)
    }

    override fun putAll(from: Map<out BlockState, V>) {
        for ((k, v) in from) {
            val hash = hash(k)
            keysMap[hash] = k
            valuesMap[hash] = v
        }
    }

    override fun remove(key: BlockState): V? {
        val hash = hash(key)
        keysMap.remove(hash)
        return valuesMap.remove(hash)
    }

    /**
     * An entry in a [BlockStateMap].
     */
    data class Entry<V>(override var key: BlockState, override var value: V) : MutableMap.MutableEntry<BlockState, V> {
        override fun setValue(newValue: V): V {
            val old = value
            value = newValue
            return old
        }
    }
}